exporter/prometheusremotewriteexporter/wal.go

```diff
@@ -139,13 +156,136 @@ func (prwe *prweWAL) start(ctx context.Context) error {
+	}()
+	<-waitUntilStartedCh
+	return nil
+}
+
+// continuallyPopWALThenExport reads a prompb.WriteRequest proto encoded blob from the WAL, and moves
+// the WAL's front index forward until either the read buffer period expires or the maximum
+// buffer size is exceeded. When either of the two conditions are matched, it then exports
+// the requests to the Remote-Write endpoint, and then truncates the head of the WAL to where
+// it last read from.
+func (prwe *prweWAL) continuallyPopWALThenExport(ctx context.Context, signalStart func()) (err error) {
+	var reqL []*prompb.WriteRequest
+	defer func() {
+		// Keeping it within a closure to ensure that the later
+		// updated value of reqL is always flushed to disk.
+		if errL := prwe.exportSink(ctx, reqL); len(errL) != 0 && err == nil {
+			err = multierror.Append(nil, errL...)
+		}
+	}()
+
+	freshTimer := func() *time.Timer {
+		return time.NewTimer(prwe.walConfig.truncateFrequency())
+	}
+
+	timer := freshTimer()
+	defer func() {
+		// Added in a closure to ensure we capture the later
+		// updated value of timer when changed in the loop below.
+		timer.Stop()
+	}()
+
+	signalStart()
+
+	maxCountPerUpload := prwe.walConfig.bufferSize()
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-prwe.stopChan:
+			return nil
+		default:
+		}
+
+		req, err := prwe.readPrompbFromWAL(ctx, atomic.LoadUint64(&prwe.rWALIndex))
+		if err != nil {
+			return err
+		}
+		reqL = append(reqL, req)
+
+		// Now increment the WAL's read index.
+		atomic.AddUint64(&prwe.rWALIndex, 1)
+
+		shouldExport := false
+		select {
+		case <-timer.C:
+			shouldExport = true
+		default:
+			shouldExport = len(reqL) >= maxCountPerUpload
+		}
+
+		if !shouldExport {
+			continue
+		}
+
+		// Otherwise, it is time to export, flush and then truncate the WAL, but also to kill the timer!
+		timer.Stop()
+		timer = freshTimer()
+
+		if err := prwe.exportThenFrontTruncateWAL(ctx, reqL); err != nil {
+			return err
+		}
+		// Reset but reuse the write requests slice.
+		reqL = reqL[:0]
+	}
+}
+
 func (prwe *prweWAL) closeWAL() {
 	if prwe.wal != nil {
 		prwe.wal.Close()
 		prwe.wal = nil
 	}
 }
 
+func (prwe *prweWAL) syncAndTruncateFront() error {
+	prwe.mu.Lock()
+	defer prwe.mu.Unlock()
+
+	if prwe.wal == nil {
+		return errNilWAL
+	}
+
+	// Save all the entries that aren't yet committed, to the tail of the WAL.
+	if err := prwe.wal.Sync(); err != nil {
+		return err
+	}
+	// Truncate the WAL from the front for the entries that we already
+	// read from the WAL and had already exported.
+	if err := prwe.wal.TruncateFront(atomic.LoadUint64(&prwe.rWALIndex)); err != nil && err != wal.ErrOutOfRange {
+		return err
+	}
+	return nil
+}
+
+func (prwe *prweWAL) exportThenFrontTruncateWAL(ctx context.Context, reqL []*prompb.WriteRequest) error {
+	if len(reqL) == 0 {
+		return nil
+	}
+	if cErr := ctx.Err(); cErr != nil {
+		return nil
+	}
+
+	if errL := prwe.exportSink(ctx, reqL); len(errL) != 0 {
```

Comment:
Shouldn't the lock include the export sink? A concurrent call to this would potentially send the same "front" twice.

Reply:
I don't think locking is a concern here, but walking back from here does have me concerned that an error returned here would stop the goroutine that reads from the WAL, but not the exporter, and there doesn't appear to be a mechanism to restart reading from the WAL.

I need to spend some more time with this.